File: tests/python/test_radioss_th.py
"""OpenRadioss time-history files (``<run>T01``): both engines on a T01 that
OpenRadioss wrote for a meshio++ deck, checked against OpenRadioss's own
``th_to_csv`` reading, and on synthetic files for the layout's other branches."""

import pathlib
import shutil
import struct

import numpy as np
import pytest

import meshioplusplus
from meshioplusplus import _core
from meshioplusplus.radioss_th import _th as py_th

COLUMN = pathlib.Path(__file__).parent / "meshes" / "radioss_th" / "column"
T01 = COLUMN / "columnT01"
REFERENCE = COLUMN / "columnT01_th_to_csv.csv"


@pytest.fixture(params=["core", "python"])
def engine(request):
    if request.param == "core":
        return meshioplusplus.radioss_th
    return py_th


def _record(payload):
    return struct.pack(">I", len(payload)) + payload + struct.pack(">I", len(payload))


def _ints(*values):
    return _record(struct.pack(f">{len(values)}i", *values))


def _floats(*values):
    return _record(struct.pack(f">{len(values)}f", *values))


def _text(text, width):
    return text.ljust(width).encode()


def th_file(version=3040, outputs=2, cut=True):
    """Globals 1, 2; parts 7 (IE, KE) and 8 (none); subset 0 (KE); TH groups:
    nodes 11 and 12 (three variables), rigid body 5 (none)."""
    title = 100 if version >= 4021 else 80 if version >= 3041 else 40
    out = _record(struct.pack(">i", version) + _text("run", 80))
    out += _record(_text("Mon Mar 30 16:39:04 2026 RADIOSS", 80))
    if version > 3050:
        out += _ints(1) + _ints(title) + _floats(1.0, 1000.0, 0.001)
    out += _ints(2, 1, 0, 1, 2, 2) + _ints(1, 2)
    out += _record(
        struct.pack(">i", 7) + _text("part7", title) + struct.pack(">4i", 0, 0, 0, 2)
    )
    out += _ints(1, 2)
    out += _record(
        struct.pack(">i", 8) + _text("part8", title) + struct.pack(">4i", 0, 0, 0, 0)
    )
    out += _record(struct.pack(">i", 1) + _text("steel", title))
    out += _record(struct.pack(">5i", 0, 0, 0, 2, 1) + _text("GLOBAL MODEL", title))
    out += _ints(7, 8) + _ints(2)
    out += _record(struct.pack(">5i", 3, 0, 0, 2, 3) + _text("th_nodes", title))
    out += _record(struct.pack(">i", 11) + _text("a", title))
    out += _record(struct.pack(">i", 12) + _text("b", title))
    out += _ints(1, 2, 3)
    out += _record(struct.pack(">5i", 4, 103, 0, 1, 0) + _text("TH RBODY", title))
    out += _record(struct.pack(">i", 5) + _text("body", title))
    for k in range(outputs):
        out += _floats(0.5 * k) + _floats(10 + k, 20 + k) + _floats(1 + k, 2 + k)
        out += (
            _floats(3 + k) + _floats(*[i + 100.0 * k for i in range(6)]) + _record(b"")
        )
    if cut:
        out += _floats(0.5 * outputs) + _floats(1.0)
    return out


def test_engines_agree():
    for step in range(len(py_th.time_values(T01))):
        a = meshioplusplus.radioss_th.read(T01, time_step=step)
        b = py_th.read(T01, time_step=step)
        assert sorted(a.field_data) == sorted(b.field_data)
        for name in a.field_data:
            assert a.field_data[name].dtype == b.field_data[name].dtype
            np.testing.assert_array_equal(a.field_data[name], b.field_data[name])


def test_real_run_matches_th_to_csv(engine):
    """Every output of OpenRadioss's T01, value for value, against th_to_csv
    (which prints 7 significant digits)."""
    reference = np.loadtxt(REFERENCE, delimiter=",")
    times = engine.time_values(T01)
    assert len(times) == len(reference) == 8
    np.testing.assert_allclose(times, reference[:, 0], rtol=1e-6)
    for step, row in enumerate(reference):
        mesh = engine.read(T01, time_step=step)
        fd = mesh.field_data
        assert len(mesh.points) == 0
        # the time, the 22 global variables (codes 1 to 22) in file order
        values = [fd["meshio:time"][0]]
        values += [
            fd[f"radioss_th:global:{py_th._global_name(c)}"][0] for c in range(1, 23)
        ]
        # part 1, then the TH groups in file order: bricks, then nodes
        tail = [fd[f"radioss_th:part:1:{v}"][0] for v in ("IE", "KE", "MASS")]
        tail += list(fd["radioss_th:brick:2"].ravel())
        tail += list(fd["radioss_th:node:1"].ravel())
        np.testing.assert_allclose(values + tail, row, rtol=1e-6, atol=1e-30)
    last = engine.read(T01, time_step=-1).field_data
    np.testing.assert_array_equal(last["radioss_th:node:1:ids"], [1, 9, 12])
    np.testing.assert_array_equal(
        last["radioss_th:node:1:variables"], [1, 2, 3, 4, 5, 6]
    )
    np.testing.assert_array_equal(last["radioss_th:brick:2:ids"], [1, 2])
    # node 1 is held
    assert not last["radioss_th:node:1"][0].any()


def test_synthetic_layout(engine, tmp_path):
    path = tmp_path / "runT01"
    path.write_bytes(th_file())
    assert engine.time_values(path) == [0.0, 0.5]  # the cut third output is dropped
    fd = engine.read(path, time_step=-1).field_data
    assert fd["radioss_th:global:internal_energy"][0] == 11.0
    assert fd["radioss_th:part:7:KE"][0] == 3.0
    assert fd["radioss_th:subset:0:KE"][0] == 4.0
    assert fd["radioss_th:node:3"].shape == (2, 3)
    assert fd["radioss_th:node:3"][1, 2] == 105.0
    assert fd["radioss_th:rbody:4"].shape == (1, 0)
    assert "radioss_th:unit_factors" not in fd
    newer = tmp_path / "newT01"
    newer.write_bytes(th_file(version=4021, cut=False))
    fd = engine.read(newer).field_data
    np.testing.assert_allclose(
        fd["radioss_th:unit_factors"], [1.0, 1000.0, 0.001], rtol=1e-7
    )
    assert fd["radioss_th:part:7:KE"][0] == 2.0


def test_narrowed(engine):
    mesh = engine.read(T01, arrays=["radioss_th:part:1:KE"])
    assert set(mesh.field_data) == {"meshio:time", "radioss_th:part:1:KE"}
    bare = engine.read(T01, points_only=True)
    assert set(bare.field_data) == {"meshio:time"}


def test_found_by_name_and_by_content(tmp_path):
    assert meshioplusplus.sniff_format(T01) == "radioss_th"
    renamed = tmp_path / "history.bin"
    shutil.copy(T01, renamed)
    assert meshioplusplus.sniff_format(renamed) == "radioss_th"
    assert _core.sniff_format(str(renamed)) == "radioss_th"
    mesh = meshioplusplus.read(T01, time_step=3)
    assert mesh.field_data["meshio:time"][0] == pytest.approx(
        meshioplusplus.radioss_th.time_values(T01)[3]
    )
    assert meshioplusplus.read_metadata(T01)["time_values"] == pytest.approx(
        meshioplusplus.radioss_th.time_values(T01)
    )
    steps = list(meshioplusplus.read_sequence(T01))
    assert len(steps) == 8


def test_refusals(engine, tmp_path):
    junk = tmp_path / "junkT01"
    junk.write_bytes(b"\x00" * 64)
    with pytest.raises(meshioplusplus.ReadError):
        engine.read(junk)
    with pytest.raises(meshioplusplus.ReadError, match="out of range"):
        engine.read(T01, time_step=8)
    cut = tmp_path / "cutT01"
    cut.write_bytes(th_file(outputs=0))
    with pytest.raises(meshioplusplus.ReadError, match="no complete output"):
        engine.read(cut)

File: tests/cpp/test_radioss_th.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//
// System includes
#include <cstdint>
#include <cstring>
#include <filesystem>
#include <fstream>
#include <random>
#include <string>
#include <vector>

// External includes
#include <gtest/gtest.h>

// Project includes
#include "mesh_fixtures.hpp"
#include "meshioplusplus/exceptions.hpp"
#include "meshioplusplus/formats/radioss_th.hpp"
#include "meshioplusplus/operations/sniff.hpp"
#include "meshioplusplus/registry.hpp"

namespace {

using meshioplusplus::Mesh;
using meshioplusplus::ReadError;
using meshioplusplus::ReadOptions;

// A big-endian Fortran unformatted file, record by record.
class Th {
public:
    // Opens a record; Close() ends it.
    Th& Open() {
        mStart = mBytes.size();
        Word(0);
        return *this;
    }
    Th& Close() {
        const std::uint32_t n = static_cast<std::uint32_t>(mBytes.size() - mStart - 4);
        for (int i = 0; i < 4; ++i)
            mBytes[mStart + i] = static_cast<char>(n >> (24 - 8 * i));
        Word(n);
        return *this;
    }
    Th& Word(std::uint32_t V) {
        for (int i = 0; i < 4; ++i)
            mBytes.push_back(static_cast<char>(V >> (24 - 8 * i)));
        return *this;
    }
    Th& Int(std::int32_t V) { return Word(static_cast<std::uint32_t>(V)); }
    Th& Float(float V) {
        std::uint32_t bits = 0;
        std::memcpy(&bits, &V, 4);
        return Word(bits);
    }
    Th& Text(const std::string& rText, std::size_t Width) {
        std::string t = rText;
        t.resize(Width, ' ');
        mBytes += t;
        return *this;
    }
    Th& Ints(const std::vector<std::int32_t>& rValues) {
        Open();
        for (const std::int32_t v : rValues)
            Int(v);
        return Close();
    }
    std::string mBytes;

private:
    std::size_t mStart = 0;
};

// Globals 1, 2; parts 7 (IE, KE) and 8 (none); subset 0 (KE); TH groups: nodes
// 11 and 12 (3 variables), rigid body 5 (none). Two outputs (t = 0, 0.5) and a
// third cut short. Version 4021 adds the unit factors and 100-character titles.
std::string th_file(std::int32_t Version = 3040) {
    const std::size_t title = Version >= 4021 ? 100 : 40;
    Th t;
    t.Open().Int(Version).Text("run", 80).Close();
    t.Open().Text("Mon Mar 30 16:39:04 2026 RADIOSS", 80).Close();
    if (Version > 3050) {
        t.Ints({1});
        t.Ints({static_cast<std::int32_t>(title)});
        t.Open().Float(1.0f).Float(1000.0f).Float(0.001f).Close();
    }
    t.Ints({2, 1, 0, 1, 2, 2});  // parts, materials, properties, subsets, groups, globals
    t.Ints({1, 2});
    t.Open().Int(7).Text("part7", title).Int(0).Int(0).Int(0).Int(2).Close();
    t.Ints({1, 2});
    t.Open().Int(8).Text("part8", title).Int(0).Int(0).Int(0).Int(0).Close();
    t.Open().Int(1).Text("steel", title).Close();
    t.Open().Int(0).Int(0).Int(0).Int(2).Int(1).Text("GLOBAL MODEL", title).Close();
    t.Ints({7, 8});
    t.Ints({2});
    t.Open().Int(3).Int(0).Int(0).Int(2).Int(3).Text("th_nodes", title).Close();
    t.Open().Int(11).Text("a", title).Close();
    t.Open().Int(12).Text("b", title).Close();
    t.Ints({1, 2, 3});
    t.Open().Int(4).Int(103).Int(0).Int(1).Int(0).Text("TH RBODY", title).Close();
    t.Open().Int(5).Text("body", title).Close();
    for (int k = 0; k < 3; ++k) {
        const float s = static_cast<float>(k);
        t.Open().Float(0.5f * s).Close();
        if (k == 2)
            break;  // cut short
        t.Open().Float(10.0f + s).Float(20.0f + s).Close();
        t.Open().Float(1.0f + s).Float(2.0f + s).Close();
        t.Open().Float(3.0f + s).Close();
        t.Open();
        for (int i = 0; i < 6; ++i)
            t.Float(static_cast<float>(i) + 100.0f * s);
        t.Close();
        t.Open().Close();
    }
    return t.mBytes;
}

std::string write_file(const std::string& rBytes, const std::string& rName = "runT01") {
    const std::filesystem::path dir(
        mt::temp_path("_radioss_th_" + std::to_string(std::random_device{}())));
    std::filesystem::create_directories(dir);
    std::ofstream((dir / rName).string(), std::ios::binary) << rBytes;
    return (dir / rName).string();
}

ReadOptions step(int Step) {
    ReadOptions o;
    o.mTimeStep = Step;
    return o;
}

double scalar(const Mesh& rMesh, const std::string& rName) {
    return rMesh.FieldData(rName).As<double>()[0];
}

}  // namespace

TEST(RadiossTh, OutputsGlobalsPartsSubsetsAndGroups) {
    const std::string path = write_file(th_file());
    EXPECT_EQ(meshioplusplus::radioss_th_time_values(path), (std::vector<double>{0.0, 0.5}));
    const Mesh last = meshioplusplus::read_radioss_th(path, step(-1));
    EXPECT_EQ(last.NumPoints(), 0u);
    EXPECT_EQ(scalar(last, "meshio:time"), 0.5);
    EXPECT_EQ(scalar(last, "radioss_th:global:internal_energy"), 11.0);
    EXPECT_EQ(scalar(last, "radioss_th:global:kinetic_energy"), 21.0);
    EXPECT_EQ(scalar(last, "radioss_th:part:7:IE"), 2.0);
    EXPECT_EQ(scalar(last, "radioss_th:part:7:KE"), 3.0);
    EXPECT_EQ(scalar(last, "radioss_th:subset:0:KE"), 4.0);
    const auto& nodes = last.FieldData("radioss_th:node:3");
    ASSERT_EQ(nodes.Shape(), (std::vector<std::size_t>{2, 3}));
    EXPECT_EQ(nodes.As<double>()[5], 105.0);
    EXPECT_EQ(last.FieldData("radioss_th:node:3:ids").As<std::int64_t>()[1], 12);
    EXPECT_EQ(last.FieldData("radioss_th:node:3:variables").As<std::int64_t>()[2], 3);
    EXPECT_EQ(last.FieldData("radioss_th:rbody:4").Shape(), (std::vector<std::size_t>{1, 0}));
    EXPECT_FALSE(last.HasFieldData("radioss_th:unit_factors"));
    EXPECT_THROW(meshioplusplus::read_radioss_th(path, step(2)), ReadError);
}

TEST(RadiossTh, NewerVersionUnitFactorsAndLongTitles) {
    const std::string path = write_file(th_file(4021));
    const Mesh first = meshioplusplus::read_radioss_th(path);
    EXPECT_EQ(scalar(first, "radioss_th:part:7:KE"), 2.0);
    const auto& f = first.FieldData("radioss_th:unit_factors");
    EXPECT_EQ(f.As<double>()[1], 1000.0);
}

TEST(RadiossTh, ResolvedSniffedNarrowedAndRefused) {
    const std::string path = write_file(th_file());
    EXPECT_EQ(meshioplusplus::resolve_format(path, ""), "radioss_th");
    EXPECT_EQ(meshioplusplus::sniff_format(write_file(th_file(), "history.bin")), "radioss_th");
    EXPECT_TRUE(meshioplusplus::is_radioss_th_filename("/a/crashT02"));
    EXPECT_FALSE(meshioplusplus::is_radioss_th_filename("/a/crashT02.csv"));
    EXPECT_FALSE(meshioplusplus::is_radioss_th_filename("/a/crashA001"));
    const auto meta = meshioplusplus::registry_read_metadata(path, "radioss_th", ReadOptions{});
    EXPECT_EQ(meta.mTimeValues.size(), 2u);
    ReadOptions only;
    only.mDataArrays = std::vector<std::string>{"radioss_th:part:7:IE"};
    const Mesh narrowed = meshioplusplus::read_radioss_th(path, only);
    EXPECT_TRUE(narrowed.HasFieldData("radioss_th:part:7:IE"));
    EXPECT_FALSE(narrowed.HasFieldData("radioss_th:node:3"));
    EXPECT_THROW(meshioplusplus::read_radioss_th(write_file(std::string(64, '\0'))), ReadError);
    // the descriptions and one time: no complete output
    const std::string text = th_file();
    EXPECT_THROW(meshioplusplus::read_radioss_th(write_file(text.substr(0, 848))), ReadError);
}
